src/Input/optionparser.cpp

```diff
@@ -471,25 +471,25 @@ void OptionParser::parseQualOption(const string& s2, const string& s3,
         network->options.setOption(Options::QUAL_UNITS, Options::MGL);
     }
 
-    if ( !s3.empty() )
+    if (network->option(Options::QUAL_TYPE) == Options::TRACE)
     {
-        if ( network->option(Options::QUAL_TYPE) == Options::TRACE )
-        {
-            int nodeIndex = network->indexOf(Element::NODE, s3);
-            if (i < 0) throw InputError(InputError::UNDEFINED_OBJECT, s3);
-            network->options.setOption(Options::TRACE_NODE, nodeIndex);
-            network->options.setOption(Options::TRACE_NODE_NAME, s3);
-        }
-        if ( network->option(Options::QUAL_TYPE) == Options::CHEM )
-        {
-            string s3U = Utilities::upperCase(s3);
-            if ( s3U.compare("MG/L") == 0 )
-                network->options.setOption(Options::QUAL_UNITS, Options::MGL);
-            else if ( s3U.compare("UG/L") == 0 )
-                network->options.setOption(Options::QUAL_UNITS, Options::UGL);
-            else throw InputError(InputError::INVALID_KEYWORD, s3);
-        }
-     }
+        if (s3.empty()) throw InputError(InputError::UNDEFINED_OBJECT, s3);
+        int nodeIndex = network->indexOf(Element::NODE, s3);
+        if (nodeIndex < 0) throw InputError(InputError::UNDEFINED_OBJECT, s3);
```

Comment:
it should be a typo to use`if (i < 0)`. use 'nodeIndex' to check the third parameter "node id" exists or not

Reply:
I agree with the change regarding the TRACE node. However, for the CHEM quality type, the use of a third token to contain the chemical units is optional. The default is MG/L which is set at line 464 if the keyword CHEMICAL is used or at line 471 if a specific chemical name is used. Therefore you don't want to throw an input error if the third token is empty.